### ISSUE TITLE
ERROR:  duplicate key value violates unique constraint "tablename_pkey"

### ISSUE THREAD

**Issue by lospejos** [dimitri/pgcopydb#319]
Using `pgcopydb` version `0.11.61.g6371ee0` with last commits:

```
* 6371ee0 2023-06-13 | Remove unnecessary reading for commit_lsn after reaching startpos. (#318) (HEAD -> main, origin/main, origin/HEAD) [Shubham Dhama]
* 91f0c94 2023-06-09 | Add table column list to schema.json. (#315) [Dimitri Fontaine]
* d447a84 2023-06-08 | Fix schema queries for exclusion filters. (#314) [Dimitri Fontaine]
* 76346e2 2023-06-07 | When applying CDC (logical replication), set role to replica. (#308) [Dimitri Fontaine]
* 090a1ca 2023-06-05 | Add unit testing coverage for "generated as identity" columns. (#309) [Dimitri Fontaine]
* 9a5cb04 2023-06-05 | Assorted set of small fixes. (#307) [Dimitri Fontaine]
* 5911755 2023-06-05 | Fix file rotation issue during streaming. (#298) [Shubham Dhama]
* cda0e34 2023-06-02 | Use JSON format for work-directory summary files. (#300) [Dimitri Fontaine]
* af554b5 2023-06-02 | Switch from semaphores to message queue to share workload. (#305) [Dimitri Fontaine]
* f060895 2023-06-02 | Avoid testing with postgis_topology. (#306) [Dimitri Fontaine]
```

Started data migration:
```
# pgcopydb clone --follow --table-jobs 8 --index-jobs 8 --restart --no-acl --no-owner --no-role-passwords --no-comments --fail-fast --filter /opt/pgcopydb_vanilla_filter.ini --plugin wal2json --trace
```

Initial load was ok, database replication started also ok.
Before data changes on source database, I compared data in source and target databases, and it was ok (compared values of last records by id for 15 important tables).

After updating data in source database (insert and update statements), error appeared and data migration process was interrupted:

```
2023-06-13 09:50:26 1564315 ERROR  pgsql.c:1892              Postgres result status is PGRES_FATAL_ERROR
2023-06-13 09:50:26 1564315 ERROR  pgsql.c:1543              [TARGET] ERROR:  duplicate key value violates unique constraint "tablename_pkey"
2023-06-13 09:50:26 1564315 ERROR  pgsql.c:1543              [TARGET] DETAIL:  Key (id)=(2946820) already exists.
2023-06-13 09:50:26 1564315 ERROR  pgsql.c:1548              SQL query: INSERT INTO "s"."tablename" ("id", "extid", "created", "updated", <skipped columns list>) overriding system value VALUES (2.94682e+06, NULL, '2023-06-13 09:46:27.634301', '2023-06-13 09:46:27.634301', <skipped column values>)
2023-06-13 09:50:26 1564315 ERROR  pgsql.c:1549              SQL params:
```

Some notes:

1) Each next run different tables appeared in log file, but every time it was complain about `_pkey` value;
2) The type of `id` column is `bigserial` (`bigint` with sequences), and in log I see a floating point value (f.e. `2.94682e+06`). I'm not sure is it a logging issue (IRL it was a bigint value but logged as floating point for some reason, or it was really attempt to set a floating point value to a `bigint` column). On target database `id` column type is `bigserial` (which is expected and ok);
3) On my small local database such errors not appeared even after inserting and updating data on source database. Errors appeared on our big near-production database.
4) On my local databases (where no errors appeared) I migrated PG14 -> PG14 (the same version of PG), but on big near-prod database I migrated PG10 (Azure Single) -> PG14 (Azure Flexible). I'm not sure could it be the error cause or not.

**Comment by dimitri** [comment]
I think when using wal2json we need to use the option `numeric-data-types-as-string` (which defaults to false). See docs at https://github.com/eulerto/wal2json. Trouble is I'm not sure if that option is widely available yet. Can you try out and report back?

**Comment by shubhamdhama** [comment]
See #127 for more information.

**Comment by dimitri** [comment]
See also https://github.com/eulerto/wal2json/issues/265#issuecomment-1424642787 for wal2json side of things.

**Comment by lospejos** [comment]
Thanks for responses, I have a couple thoughts though.

1) Unfortunately I cannot set `numeric-data-types-as-string`  option on my servers since they are Azure PostgreSQL where there is no possibility to edit `postgresql.conf`. I checked the `Server Parameters` and there are no such server parameter. I'll try to figure out can I set it using `az` CLI, but not sure;

2) The error says, that new record id value already was updated/set on target database somehow, and the seconds attempt to insert record with the same id value leads to an error. This means, the first time somehow data was correctly parsed and executed on target database. And this means, IMHO, that above-mentioned cause (`wasl2json` plugin) may not be the real cause of this issue. And even if I somehow set `numeric-data-types-as-string = true` on source database, this issue still happen.

**Comment by dimitri** [comment]
Hi @lospejos ; we're talking about a logical decoding option, that needs to be passed in when creating the replication slot with the plugin. See in the code at the following place:

https://github.com/dimitri/pgcopydb/blob/ab9090a8643129f4f8f849abe6bcaf5fec2e1822/src/bin/pgcopydb/ld_stream.c#L94-L112

I am not sure about the idea in your second point though. It's not about parsing the value, more about what value has been given by wal2json and the possible precision loss there. Basically javascript has 48 bits numbers and Postgres has 64 bits numbers, if I remember correctly...

**Comment by lospejos** [comment]
@dimitri Thanks for your reply!
Regarding this
> we're talking about a logical decoding option, that needs to be passed in when creating the replication slot with the plugin

Cannot get what option/parameter should I set in `pgcopydb clone --follow --plugin wal2json ...`, found nothing in https://pgcopydb.readthedocs.io/en/latest/ref/pgcopydb_follow.html#pgcopydb-follow about it.
Could you please give an example about how to set this option in command line?

Thanks!

**Comment by dimitri** [comment]
That's the trick, it's not possible at the moment, it's source-code only. You can edit the sources to make it happen, though (pgcopydb is Open Source after all), and you need to make sure the version of wal2json that you're dealing with has the option already, knowing that no such version has been released yet in the public (so build from sources also there).

**Comment by ulo23k** [comment]
@dimitri I ran into the same issue and I followed the same instructions as you mentioned in the above thread by updating the source code and I still got the same error message

`duplicate key value violates unique constraint "dream_pony_pkey"
2023-06-27 16:34:45 29433 ERROR  pgsql.c:1529              [TARGET] DETAIL:  Key (id)=(1657560) already exists.
2023-06-27 16:34:45 29430 ERROR  follow.c:808              Subprocess catchup with pid 29433 has exited with error code 12
2023-06-27 16:34:45 29431 INFO   pgsql.c:4267              Reported write_lsn 24FA/82017068, flush_lsn 24FA/82017068, replay_lsn 24F9/30000278
2023-06-27 16:34:45 29431 INFO   ld_stream.c:442           Streamed up to write_lsn 24FA/82017068, flush_lsn 24FA/82017068, stopping
2023-06-27 16:34:46 29430 ERROR  follow.c:540              Some sub-process exited with errors, see above for details
2023-06-27 16:34:46 29424 ERROR  cli_clone_follow.c:697    follow process 29430 has terminated [12]
2023-06-27 16:34:46 29424 INFO   follow.c:110              follow_reset_sequences: host=dreamer-db.cdnmzeaj1lfr.us-east-2.rds.amazonaws.com port=5432 dbname=dreamer user=master
2023-06-27 16:34:46 29424 INFO   sequences.c:39            Fetching information for 2 sequences
2023-06-27 16:34:46 29424 INFO   sequences.c:168           Reset sequences values on the target database`

When it was trying to replay this SQL:

`2023-06-27 16:34:45 UTC:10.69.31.241(41088):master@dreamer:[14314]:STATEMENT: INSERT INTO "public"."dream_pony" ("id", "dream_id", "dream_number", "source_code", "default") overriding system value VALUES (1.65756e+06, 944589, E'0927989', E'CS', 't')
`
**FYI**: both my source and target instances are RDS instances.

**Comment by dimitri** [comment]
The test_decoding plugin support has been improved in the meantime. Could you run some tests with using it and report success or problems along the way?

**Comment by lospejos** [comment]
@dimitri 
You wrote:
_Could you run some tests with using it and report success or problems along the way?_

Does it mean I should use it like this:
```
pgcopydb clone --follow ... --plugin test_decoding
```
?

**Comment by lospejos** [comment]
I've started another data replication session on the same source and target databases, but removed `--plugin wal2json` (AFAIK this should default plugin to `test_decoding` as I see it from documentation).

Than after initial data was copied, I updated source DB, data migration process still working (no crash like before using `wal2json` plugin). No data on target database was updated (this is expected since `test_decoding` does not update data AFAIK).

Not sure what was the target of this test run.

**Comment by lospejos** [comment]
I performed additional testing on the same source (Azure Single Server PostgreSQL version 10.23) and target (Azure Flexible Server PostgreSQL version 14.7) using a simple schema and simple 2 tables with primary keys, foreign keys etc.

Initial data migration was completed without errors, streaming replication was active (used plugin `wal2json`). Then I inserted some data on source database, just like in my previous tests with large database (prod copy), and this time data was successfully migrated to a destination database.

I also tested it on large `id` values (assuming that large values may cause this issue, like `2950436` f.e. interpreted as `2.950436e+06`), but these tests also completed without errors, data with large `id` values successfully migrated with streaming replication.

Probably in our prod-copy database we have something, that causes this error. We have plenty triggers and trigger functions, also some external services (applications) also update data. I'm not sure it this may be the cause.
